numpy/core/shape_base.py

```diff
@@ -423,32 +434,112 @@ def format_index(index):
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
+    Produce the slices that match the destination each array would end up
+    along axis.
+
+    For the following situation::
 
-    return block_recursion(arrays)
+        ret = concatenate([a, b, c], axis=0)
+        sl_a, sl_b, sl_c = concatenate_slices([a, b, c], axis=0)
+
+    The result will be that ``ret[sl_a] == a``, ``ret[sl_b] == b``, and
+    ``ret[sl_c] == c``.
+    Similar properties hold for other values of ``axis``.
+
+    See Also
+    --------
+        concatenate : Join a sequence of arrays together.
+
+    """
+    offsets = (0,) + tuple(_accumulate(shape[axis] for shape in shapes[:-1]))
+    return [slice(offset, offset + shape[axis])
+            for offset, shape in zip(offsets, shapes)]
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Form required information to block the arrays using a flat iterator.
+
+    Returns
+    -------
+    shape: tuple of ints
+        the resulting shape of the final array
+    region_slices: iterable of indices
+        slices into the full array required for assignment
+    region_data: iterable of ndarray
+        the data to assign to each slice of the full array
+
+    """
+    if depth < list_ndim:
+        axis = result_ndim - list_ndim + depth
+        info = [_block_info_recursion(arr, list_ndim, result_ndim, depth + 1)
+                for arr in arrays]
+        shapes, slices, arrays = zip(*info)
+        # Compute the resulting shape
+        shape = _concatenate_shapes(shapes, axis)
+
+        # `slices` and `arrays` contain lists that have the information
+        # from the inner lists provided to the concatenation function.
+        # To create the correct offset, one needs to prepend the appropriate
+        # offset to each of them. Each list (containing one or more arrays)
+        # will require the matching offset (computed above with accumulate)
+        slice_prefixes = _concatenate_shapes_as_slices(shapes, axis)
+        # Prepend the slice prefix and flatten the slices
+        slices = [(slice_prefix,) + the_slice
+                  for slice_prefix, inner_slices in zip(slice_prefixes, slices)
+                  for the_slice in inner_slices]
+        # Flatted the arrays
+        arrays = [arr
+                  for inner_array in arrays
+                  for arr in inner_array]
```

Comment:
Using a simple benchmark calling on `[[1]*100]*100`, I find that using `chain.from_iterable` is 3x faster than the list comprehension you have here (and 6x faster than a generator expression:
```
In [9]: %timeit list(itertools.chain.from_iterable(data))
117 µs ± 1.49 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [10]: %timeit [x for xs in data for x in xs]
326 µs ± 11.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [11]: %timeit list(x for xs in data for x in xs)
708 µs ± 39.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
As long as what you have here is faster than what was here before though, I'm happy to leave the microoptimizations to a later PR

Reply:
No way, we already hit comment 100, might as well get to 200.